Feature: transfer TLD

hbarcelos opened the pull request:
Implements [JP-799](https://janusproj.atlassian.net/secure/RapidBoard.jspa?rapidView=7&modal=detail&selectedIssue=JP-799) and [JP-803](https://janusproj.atlassian.net/secure/RapidBoard.jspa?rapidView=7&modal=detail&selectedIssue=JP-803):

- [x] Fixes ownership transfer for TLDs to update the reversed index (`owner -> TLD`).
     - [x] Added regression test case.
- [x] Implements ownership transfer for Domains.

Commit:
feature: drafted first transfer TLD implementaiton

Commit:
chore: Updated core package to use the new smartcontract structure

Commit:
feature: Implemented core transfer TLD

Commit:
feature: Added inital structure of transfer TLD in dapp

Commit:
refactor: Improved transfer TLD implementation in dapp

Commit:
bugfix: Fixed transfer TLD and transfer domain behavior

Commit:
test: Added regression tests for edge cases of transfer TLD

Commit:
bugfix: Fixed domain and TLD list reload after a successful transfer

rodrigorsdev commented on packages/smartcontract/contracts/JanusNameService.sol:
```diff
@@ -26,21 +27,22 @@ contract JanusNameService {
+        uint ownerIndex;
     }
 
-    TopDomain[] private topDomains;
```
why did you change the topDomains array and mappings to public?

hbarcelos replied:
It helped me debug the problem we had with updating the indexes for ownership transfer.
Furthermore, I added some regression tests that depend on this property being public.

rodrigorsdev replied:
yeah, i forgot about the tests.
Make sense.

rodrigorsdev commented on packages/smartcontract/contracts/JanusNameService.sol:
```diff
@@ -26,21 +27,22 @@ contract JanusNameService {
+    mapping(bytes32 => uint) public topDomainsHashMap;
+    mapping(address => uint[]) public topDomainsOwnerMap;
 
-    Domain[] private domains;
```
why did you change the domains array and mappings to public?

hbarcelos replied:
Same reason as above.

rodrigorsdev replied:
yeah, i forgot about the tests.
Make sense.

hbarcelos commented on packages/smartcontract/test/changeDomainOwnership.test.js:
```diff
@@ -0,0 +1,145 @@
+      { from: ownerAddress }
+    );
+
+    const [transfered, kept] = await Promise.all([
+      // The index 0 is already taken by the contract
+      contractInstance.domains(1),
+      contractInstance.domains(2),
+    ]);
```
@rodrigorsdev here's the regression test I was talking about.

Commit:
refactor: Started to normalize TLD entity name

Commit:
chore: Improved contract build information handling

SmartContract build directory is now no checked in source control. Using a deep-merge algorithm ensures no data will be lost in contracts.json file

rodrigorsdev reviewed (approved):
ok, nice job
